submissions/nadsatt/post-oop/index.js

```diff
@@ -0,0 +1,152 @@
+   - FOR L: as i see, there is no cases where subclass can't be used instead of its superclass
+            for example, in 'addFriends' method of 'FriendsManager' class instance of any subclass can be used
+   - FOR I: as i see, there is no unused/unrelevant superclass' methods/props in subclasses
+   - FOR D: -
+*/ 
+
+class Inhabitant {
+   constructor(species, saying, gender, name, friends = []){
+      this.species = species;
+      this.saying = saying;
+      this.gender = gender;
+      this.name = name;
+      this.friends = friends;
+   }
+
+   toString(){
+      return `species: ${this.species}, saying: ${this.saying}, gender: ${this.gender}, name: ${this.name}, friends: ${new FriendsManager(this).getFriends()}`;
+   }
+}
+
+class FriendsManager {
+   constructor(inhabitant){
+      this.inhabitant = inhabitant;
+   }
+
+   addFriends(...friends){
+      this.inhabitant.friends.push(...friends);
+   }
+
+   getFriends(){
+      return this.inhabitant.friends.map(friend => friend.name).join(' ') || 'no friends';
+   }
+}
+
+class Primate extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.hands = 2;
+      this.legs = 2;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}, hands: ${this.hands}`;
+   }
+}
+
+class Human extends Primate {
+   constructor(saying, gender, name, ...friends){
+      super('human', saying, gender, name, ...friends);
+   }
+}
+
+class Man extends Human {
+   constructor(name, ...friends){
+      super('i\'m a man!', 'male', name, ...friends);
+   }
+}
+
+class Woman extends Human {
+   constructor(name, ...friends){
+      super('i\'m a woman!', 'female', name, ...friends);
+   }
+}
+
+class Predator extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.legs = 4;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}`;
+   }
+}
+
+class Dog extends Predator {
+   constructor(gender, name, ...friends){
+      super('dog', 'woof-woof!', gender, name, ...friends);
+   }
+}
+
+class Cat extends Predator {
+   constructor(gender, name, ...friends){
+      super('cat', 'meov-meov!', gender, name, ...friends);
+   }
+}
+
+class CatWoman extends Cat {
+   constructor(name, ...friends){
+      super('female', name, ...friends);
+      this.species = 'human';
+      this.hands = 2;
+      this.legs = 2;
+   }
+
+   toString(){
+      return super.toString() + `, hands: ${this.hands}`;
+   }
+}
+
+const cat1 = new Cat('female', 'Kitty');
+const dog1 = new Dog('male', 'Bob', cat1);
+const man1 = new Man('Nick');
+const woman1 = new Woman('Lucy');
+const woman2 = new Human('i\'m a special woman!', 'female', 'Linda');
+const catWoman1 = new CatWoman('Kate', man1, woman2);
```

Comment:
**DO NOT FIX**

**Bonus**
You have sufficient skills and knowledge to comprehend the foolowing fragment.

```
const inhabitants = [[Dog, 1], [Cat, 'a', 'b'], [Human, 7, 8]]
  .map(([className, ...args]) => new className(...args))
```
This however poses a problem: how do we then nicely refer objects to build ties (friends relations) between them?
`inhabitants[1].addFriends(inhabitants[2])` doesn't look really nice.
Well, one approach possible is to employ `Array.find` to find (e.g. by name property) both 'host' friend and a friend to add.